### PR TITLE
HackStudio+sed: Improve the templates (and accidentally improve sed)

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
There's a whole yakstack of changes here:
- Implement -i for sed for in-place file editing in the GNU version (which is non-trivial if you disregard the rather easy file suffix case).
- Remove the cpp-gui template since it's basically the same as serenity-application. Some nice window parameters are carried over.
- Modernize and expand other templates, utilizing the new sed capabilities.

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
cc @IdanHo

**Comment by nico** [comment]
Should probably mention in pr desc and commit text that this implements the GNU variant if the flag, not the bsd one, and why ("more convenient to use", i suppose)

**Comment by nico** [comment]
...but GNU sed also allows an optional suffix arg to `-i`, it's just joined instead of separate of given.

**Comment by kleinesfilmroellchen** [comment]
> Should probably mention in pr desc and commit text that this implements the GNU variant if the flag, not the bsd one, and why ("more convenient to use", i suppose)

Interesting; I did actually have a look at the OpenBSD implementation to see how they do in place editing, but it was too confusing so I came up with my own solution. I don't even know what BSD does then, since the GNU behavior is pretty intuitive to me, even if it's a little tricky to implement right.

File: Base/res/devel/templates/cpp-library.postcreate [hunk 1]
```diff
@@ -4,6 +4,10 @@
 # $2: Project full path
 # $3: Project name, namespace safe
 
+echo $2
```
Comment (alimpfard) [inline Base/res/devel/templates/cpp-library.postcreate:7]: Debug print?

File: Base/res/devel/templates/cpp-library.postcreate [hunk 2]
```diff
@@ -4,6 +4,10 @@
...
+
+sed -i s/'\$3'/"$3/g" $2/Class1.h $2/Class1.cpp
```
Comment (alimpfard) [inline Base/res/devel/templates/cpp-library.postcreate:9]: let's make this a bit less weird looking:
```suggestion
sed -i 's/\$3/'$3/g $2/Class1.h $2/Class1.cpp
```
Or really,
```suggestion
sed -i "s/\\\$3/$3/g" $2/Class1.h $2/Class1.cpp
```

File: Base/res/devel/templates/cpp-basic/main.cpp [hunk 3]
```diff
@@ -1,9 +1,25 @@
 #include <AK/Format.h>
+#include <AK/Optional.h>
```
Comment (GMTA) [inline Base/res/devel/templates/cpp-basic/main.cpp:2]: This is not used

File: Userland/Utilities/sed.cpp [hunk 4]
```diff
@@ -108,9 +108,11 @@ ErrorOr<int> serenity_main(Main::Arguments args)
 
     StringView command_input;
     Vector<StringView> filepaths;
+    bool in_place;
```
Comment (alimpfard) [inline Userland/Utilities/sed.cpp:111]: pls initialise this with some value (presumably `false`).

File: Base/res/devel/templates/cpp-library/Class1.cpp [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#include "Class1.h"
+#include <AK/Format.h>
+
+namespace $3 {
```
Comment (alimpfard) [inline Base/res/devel/templates/cpp-library/Class1.cpp:4]: Perhaps these could have actual names instead of `$3`?

**Review by alimpfard — COMMENTED** [review]
Mostly LGTM.
There's a conflict, and here's a couple minor comments.

**Comment by kleinesfilmroellchen** [comment]
So I need to rearchitect the entire in-place editing commit now... fun times

**Comment by kleinesfilmroellchen** [comment]
@alimpfard finally updated this PR, had to rewrite the entire in-place editing command since sed was basically entirely rewritten. However, I'm much happier with the result now since I abstracted the "independent and non-interfering reads and writes" into the new `DuplexFile` stream wrapper. LMK what you think; I will add tests to the stream class once I have verified the sed functionality.

File: Userland/Libraries/LibCore/DuplexFile.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+namespace Core {
+
+DuplexFile::DuplexFile(MaybeOwned<BufferedFile> stream)
+    : m_stream(move(stream))
+{
+}
+
+ErrorOr<NonnullOwnPtr<DuplexFile>> DuplexFile::create(MaybeOwned<BufferedFile> stream)
+{
+    size_t initial_pointer = 0;
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/DuplexFile.cpp:18]: nit: That's an offset though

File: Userland/Libraries/LibCore/DuplexFile.cpp [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    if (auto const result = stream->tell(); !result.is_error())
+        initial_pointer = result.value();
+
+    auto file = TRY(adopt_nonnull_own_or_enomem(new (nothrow) DuplexFile(move(stream))));
+    file->m_read_pointer = file->m_write_pointer = initial_pointer;
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/DuplexFile.cpp:23]: nit: Could you split this into two separate assignments?

File: Tests/LibCore/TestLibCoreStream.cpp [hunk 8]
```diff
@@ -612,3 +613,38 @@ TEST_CASE(buffered_tcp_socket_read)
...
+constexpr auto start_of_replacement_text = "This line of text is much much longer than what we read previously. Still, it does not overwrite anything we haven't read yet."sv;
+
+TEST_CASE(duplex_file)
+{
+    constexpr auto filename = "/tmp/duplex_file"sv;
+    auto duplex_file_creation = Core::File::open(filename, Core::File::OpenMode::Write).release_value();
+    EXPECT(!duplex_file_creation->write_until_depleted(duplexed_text.bytes()).is_error());
+    duplex_file_creation->close();
+    {
+        auto read_buffer = ByteBuffer::create_uninitialized(KiB).release_value();
```
Comment (alimpfard) [inline Tests/LibCore/TestLibCoreStream.cpp:627]: I can haz `1 * KiB`?

**Review by alimpfard — COMMENTED** [review]
I like this iteration a lot better, the abstraction seems good too.
Just a couple nits and we can :Marge:

**Comment by alimpfard** [comment]
So I spent some time checking what `sed` on linux does with inplace editing:
- The command's execution is atomic, if the process is terminated early, the file isn't left in a half-edited state (your implementation is strictly worse here)
- Linux `sed` reads and writes in 4096b chunks (might be different in different situations), I'm not quite sure what we're doing here, but your PR doesn't seem to be touching that regardless.

The atomic editing is implemented the _simple_ way: make a temp file, write edits there, rename() at the end.

On a sidenote about `DuplexFile`, @ADKaster brought up the question "is this generally useful outside sed?", and I think the answer is "no", I'm not even sure if it's needed in sed due to its non-atomic editing behaviour.

TLDR: I think we're better off with a simple implementation that does what linux sed does.

**Comment by kleinesfilmroellchen** [comment]
> The atomic editing is implemented the _simple_ way: make a temp file, write edits there, rename() at the end.

Okay, I specifically wanted to avoid anything that buffers more than necessary, but it seems that GNU just doesn't care. I guess it doesn't matter that you can't sed 4GB files on a 6GB drive.

**Comment by alimpfard** [comment]
I personally prefer having atomicity to being able to sed a very large file inplace with very limited disk space.
